torchgeo/datasets/south_africa_crop_type.py

```diff
@@ -0,0 +1,248 @@
+        self,
+        root: str = "",
+        crs: CRS = CRS.from_epsg(32634),
+        classes: list[int] = list(cmap.keys()),
+        bands: list[str] = all_bands,
+        transforms: Optional[Callable[[dict[str, Tensor]], dict[str, Tensor]]] = None,
+    ) -> None:
+        """Initialize a new South Africa dataset instance.
+
+        Args:
+            root: root directory where dataset can be found
+            crs: coordinate reference system to be used
+            classes: crop type classes to be included
+            bands: the subset of bands to load
+            transforms: a function/transform that takes input sample and its target as
+                entry and returns a transformed version
+        Raises:
+            DatasetNotFoundError: If dataset is not found and *download* is False.
+        """
+        assert (
+            set(classes) <= self.cmap.keys()
+        ), f"Only the following classes are valid: {list(self.cmap.keys())}."
+        assert 0 in classes, "Classes must include the background class: 0"
+
+        self.root = root
+        self.classes = classes
+        self.ordinal_map = torch.zeros(max(self.cmap.keys()) + 1, dtype=self.dtype)
+        self.ordinal_cmap = torch.zeros((len(self.classes), 4), dtype=torch.uint8)
+
+        super().__init__(paths=root, crs=crs, bands=bands, transforms=transforms)
+
+        # Map chosen classes to ordinal numbers, all others mapped to background class
+        for v, k in enumerate(self.classes):
+            self.ordinal_map[k] = v
+            self.ordinal_cmap[v] = torch.tensor(self.cmap[k])
+
+    def __getitem__(self, query: BoundingBox) -> dict[str, Any]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            data, label, and field ids at that index
+        """
+        hits = self.index.intersection(tuple(query), objects=True)
+        filepaths = cast(list[str], [hit.object for hit in hits])
+
+        if not filepaths:
+            raise IndexError(
+                f"query: {query} not found in index with bounds: {self.bounds}"
+            )
+
+        data_list: list[Tensor] = []
+        filename_regex = re.compile(self.filename_regex, re.VERBOSE)
+
+        sample_dates_s1: dict[str, str] = {}
+        sample_dates_s2: dict[str, str] = {}
+        for band in self.bands:
+            sample_dates = sample_dates_s1 if band in self.S1_bands else sample_dates_s2
+            fields_added = []
+            band_filepaths = []
+            for filepath in filepaths:
+                filename = os.path.basename(filepath)
+                match = re.match(filename_regex, filename)
+                if match and match.groupdict()["band"] == band:
+                    field_id = match.groupdict()["field_id"]
+                    month = match.groupdict()["month"]
+                    day = match.groupdict()["day"]
+                    if field_id not in fields_added and month == "07":
```

Comment:
What's the deal with this? Seems like there are a lot of custom hacks with no comments to explain them.